File: solution/525. Contiguous Array/525.java
class Solution {
    public int findMaxLength(int[] nums) {
        HashMap<Integer, Integer> prefix_sum = new HashMap<Integer, Integer>();
        int sum = 0;
        int ans = 0;
        
        prefix_sum.put(0, -1);
        for (int i = 0; i < nums.length; i++) {
            sum += (nums[i] != 0 ? nums[i] : -1);
            if (prefix_sum.containsKey(sum)) {
                int pre_index = prefix_sum.get(sum);
                ans = Math.max(ans, i - pre_index);
            } else {
                prefix_sum.put(sum, i);
            }
        }
        
        return ans;
    }
}

File: solution/525. Contiguous Array/525.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        seen = {0: -1}
        res, presum = 0, 0
        for i in range(len(nums)):
            presum += 1 if nums[i] == 1 else -1
            if presum in seen:
                res = max(res, i - seen[presum])
            else:
                seen[presum] = i
        return res
    
